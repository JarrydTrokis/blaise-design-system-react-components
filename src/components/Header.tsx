import React, { ReactElement } from "react";

export interface Props {
    title: string
    signOutButton?: boolean
    noSave?: boolean
    signOutFunction?: () => void
}


function Header({ title, signOutButton, noSave, signOutFunction }: Props): ReactElement {
    let signOutText = "Save and sign out"
    if (noSave) {
        signOutText = "Sign out"
    }
    return (
        <>
            <header className="header header--internal">
                <div className="header__top" role="banner">
                    <div className="container">
                        <div
                            className="header__grid-top grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap ">
                            <div className="grid__col col-auto">
                                <a className="header__logo-link" href="/">
                                    <picture>
                                        {/*<source media="(max-width: 499px)" srcSet="/img/ons-logo-stacked-neg-en.svg"*/}
                                        {/*        alt="Office for National Statistics logo"/>*/}
                                        <img className="header__logo"
                                            src="https://cdn.ons.gov.uk/sdc/design-system/31.4.0/img/ons-logo-neg-en.svg"
                                            alt="Office for National Statistics logo" />
                                    </picture>
                                </a>
                            </div>
                        </div>
                    </div>
                </div>
                <div className="header__main">
                    <div className="container">
                        <div
                            className="grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap">
                            <div className="grid__col col-auto u-flex-shrink">
                                <div className="header__title">{title}</div>
                            </div>
                            {
                                (
                                    signOutButton &&
                                    <div className="grid__col col-auto u-flex-no-shrink u-d-no@xxs@m">
                                        <button
                                            id="signout-button"
                                            data-test-id="signout-button"
                                            className="btn btn--ghost u-d-no@xxs@m btn--exit"
                                            onClick={() => signOutFunction && signOutFunction()}>
                                            <span className="btn__inner">{signOutText}
                                                <svg className="svg-icon" viewBox="0 0 12 12" xmlns="http://www.w3.org/2000/svg"
                                                    focusable="false">
                                                    <path
                                                        d="M13.85,7.65l-2.5-2.5a.5.5,0,0,0-.71,0,.48.48,0,0,0-.15.36V7h-3a.5.5,0,0,0-.5.5v1a.5.5,0,0,0,.5.5h3v1.5A.49.49,0,0,0,11,11a.48.48,0,0,0,.34-.14l2.51-2.5a.49.49,0,0,0,0-.68Z"
                                                        transform="translate(-2 -2)" />
                                                    <path
                                                        d="M8.5,14h-6a.5.5,0,0,1-.5-.5V2.5A.5.5,0,0,1,2.5,2h6a.5.5,0,0,1,.5.5V3a.5.5,0,0,1-.5.5h-5v9h5A.5.5,0,0,1,9,13v.5A.5.5,0,0,1,8.5,14Z"
                                                        transform="translate(-2 -2)" />
                                                </svg>
                                            </span>
                                        </button>
                                    </div>
                                )
                            }
                        </div>
                    </div>
                </div>
            </header>
        </>
    );
}

export default Header;
