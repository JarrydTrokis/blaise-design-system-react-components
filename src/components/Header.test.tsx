import React from "react";
import { fireEvent, render } from "@testing-library/react";
import sinon from "sinon";
import Header, { Props } from "./Header";

const testProps: Props = {
    title: "ONS Service",
    navigationLinks: [
        { label: "Home", endpoint: "/" },
        { label: "Deploy a questionnaire", endpoint: "/deploy" },
        { label: "View deployment history", endpoint: "/history" },
        { label: "Check Blaise status", endpoint: "/status" },
    ],
};

describe("Check default Header:", () => {
    it("matches Snapshot", () => {
        const wrapper = render(
            <Header title={testProps.title} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it("should render correctly", () => {
        const wrapper = render(
            <Header title={testProps.title} />
        );
        expect(wrapper).toBeDefined();
    });

    it("should render with the title displayed", () => {
        const wrapper = render(
            <Header title={testProps.title} />
        );
        expect(wrapper.getByText(testProps.title)).toBeVisible();
    });

    it("should not show the signout button by default", () => {
        const wrapper = render(
            <Header title={testProps.title} />
        );
        expect(wrapper.queryAllByText(/Save and sign out/)).toStrictEqual([]);
        expect(wrapper.queryAllByText(/Sign out/)).toStrictEqual([]);
    });

    it("should not show the navigation by default", () => {
        const wrapper = render(
            <Header title={testProps.title} />
        );
        expect(wrapper.queryByRole(/navigation/)).toStrictEqual(null);
    });
});

describe("Check Header with sign out button:", () => {
    it("matches Snapshot with sign out button", () => {
        const wrapper = render(
            <Header title={testProps.title} signOutButton signOutFunction={jest.fn()} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it("shows sign out button", () => {
        
        const wrapper = render(
                <Header title={testProps.title} signOutButton signOutFunction={jest.fn()} />
        );
        expect(wrapper.getByText(/Save and sign out/)).toBeVisible();
    });

    it("shows sign out button with special text", () => {
        const wrapper = render(
                <Header title={testProps.title} signOutButton noSave signOutFunction={jest.fn()} />
        );
        expect(wrapper.getByText(/Sign out/)).toBeDefined();
        expect(wrapper.queryAllByText(/Save and sign out/)).toStrictEqual([]);
    });

    it("passes in the sign out function correctly button", () => {
        const mockFunction = sinon.spy();
        const wrapper = render(
                <Header title={testProps.title} signOutButton signOutFunction={mockFunction} />
        );
        fireEvent.click(wrapper.getByText(/Save and sign out/));
        expect(mockFunction).toHaveProperty("callCount", 1);
    });
});

describe("Check Header with navigation bar:", () => {
    it("matches Snapshot with navigation", () => {
        const wrapper = render(
                <Header title={testProps.title} navigationLinks={testProps.navigationLinks} />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it("shows the navigation with links", () => {
        const wrapper = render(
                <Header title={testProps.title} navigationLinks={testProps.navigationLinks} />
        );
        expect(wrapper.queryByRole(/navigation/)).not.toStrictEqual(null);
        expect(wrapper.getByRole("link", { name: /Home/ })).toBeVisible();
        expect(wrapper.getByRole("link", { name: /Deploy a questionnaire/ })).toBeVisible();
        expect(wrapper.getByRole("link", { name: /View deployment history/ })).toBeVisible();
        expect(wrapper.getByRole("link", { name: /Check Blaise status/ })).toBeVisible();
    });
});
