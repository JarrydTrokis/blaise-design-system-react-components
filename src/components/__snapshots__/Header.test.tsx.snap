// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Test matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header header--internal"
      >
        <div
          class="header__top"
          role="banner"
        >
          <div
            class="container"
          >
            <div
              class="header__grid-top grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap "
            >
              <div
                class="grid__col col-auto"
              >
                <a
                  class="header__logo-link"
                  href="/"
                >
                  <picture>
                    <img
                      alt="Office for National Statistics logo"
                      class="header__logo"
                      src="https://ons-design-system.netlify.app/img/ons-logo-neg-en.svg"
                    />
                  </picture>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="header__main"
        >
          <div
            class="container"
          >
            <div
              class="grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap"
            >
              <div
                class="grid__col col-auto u-flex-shrink"
              >
                <div
                  class="header__title"
                >
                  App Title
                </div>
              </div>
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="header header--internal"
    >
      <div
        class="header__top"
        role="banner"
      >
        <div
          class="container"
        >
          <div
            class="header__grid-top grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap "
          >
            <div
              class="grid__col col-auto"
            >
              <a
                class="header__logo-link"
                href="/"
              >
                <picture>
                  <img
                    alt="Office for National Statistics logo"
                    class="header__logo"
                    src="https://ons-design-system.netlify.app/img/ons-logo-neg-en.svg"
                  />
                </picture>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="header__main"
      >
        <div
          class="container"
        >
          <div
            class="grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap"
          >
            <div
              class="grid__col col-auto u-flex-shrink"
            >
              <div
                class="header__title"
              >
                App Title
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
